Support more strategies for variable (re)initialization in `ElementSubdomainModifier`

ChengHauYang opened the pull request:
close #30627
close #29054

Add more strategies for variable (re)initialization in `ElementSubdomainModifier`

The `ElementSubdomainModifier` dynamically changes subdomains. When elements change subdomain, we should allow user to specify if/how variables should be (re)initialized.

Currently, we only support initial condition projection as the initialization strategy. This PR adds three strategies for variable (re)initialization:
- `POLYNOMIAL_NEIGHBOR`
- `POLYNOMIAL_NEARBY`
- `POLYNOMIAL_WHOLE`

All three strategies employ the same algorithm:
1. Identify the updated subdomain on which a variable should be (re)initialized.
2. Construct a patch of elements.
3. Fit a polynomial over the element patch.
4. Project the extrapolation of the fitted polynomial onto the updated subdomain.

At step 3, the polynomial is fitted in the least-squares sense against variable value interpolated onto quadrature points. User has control over the polynomial degree. The current implementation hard codes the polynomial basis to be monomial -- this could be relaxed in the future.

The three strategies differ only in step 2. `POLYNOMIAL_NEIGHBOR` constructs the patch using the direct element neighbors of the updated subdomain. `POLYNOMIAL_NEARBY` performs a geometric search to include elements within a given distance. `POLYNOMIAL_WHOLE` uses the entire mesh as the patch.

ChengHauYang committed:
Support more strategies for variable (re)initialization in ElementSubdomainModifier

Add more strategies for variable (re)initialization in ElementSubdomainModifier

The ElementSubdomainModifier dynamically changes subdomains. When elements change subdomain, we should allow user to specify if/how variables should be (re)initialized.

Currently, we only support initial condition projection as the initialization strategy. This PR adds three strategies for variable (re)initialization:

- POLYNOMIAL_NEIGHBOR
- POLYNOMIAL_NEARBY
- POLYNOMIAL_WHOLE
All three strategies employ the same algorithm:

Identify the updated subdomain on which a variable should be (re)initialized.
Construct a patch of elements.
Fit a polynomial over the element patch.
Project the extrapolation of the fitted polynomial onto the updated subdomain.
At step 3, the polynomial is fitted in the least-squares sense against variable value interpolated onto quadrature points. User has control over the polynomial degree. The current implementation hard codes the polynomial basis to be monomial -- this could be relaxed in the future.

The three strategies differ only in step 2. POLYNOMIAL_NEIGHBOR constructs the patch using the direct element neighbors of the updated subdomain. POLYNOMIAL_NEARBY performs a geometric search to include elements within a given distance. POLYNOMIAL_WHOLE uses the entire mesh as the patch.

close #30627

hugary1995 commented:
@roystgnr when you get a chance, could you take a look at this PR (mostly at the concept level)? I think the PR description is fairly thorough, but basically we are trying to construct a polynomial over a patch of elements, then use the constructed polynomial to perform extrapolation. User will have control over the polynomial order.

This should be a better approach than our previous attempt which simply finds the nearest dof. I think this approach works for any function spaces.

lindsayad commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -191,13 +191,6 @@ FEProblemBase::validParams()
-  ///    be set for the whole domain
-  /// 2. _blocks.size() > 0 and no coordinate system was specified, then the whole domain will be XYZ.
-  /// 3. _blocks.size() > 0 and one coordinate system was specified, then the whole domain will be that system.
-  params.addDeprecatedParam<std::vector<SubdomainName>>(
```
If you are changing the meaning of this parameter then you need to remove the previously associated (deprecated) parameters `coord_type` and `rz_coord_axis`

ChengHauYang replied:
Hi @lindsayad, Thanks for your suggestions. @hugary1995  is going to work on removing this. Thanks!

lindsayad commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -497,9 +513,30 @@ FEProblemBase::FEProblemBase(const InputParameters & parameters)
+  checkConflict(_material_coverage_check, "material_coverage_check");
+
+  if (isParamSetByUser("block") && !_has_block_in_global)
+    mooseWarning("The block parameter is set by user in the Problem block, but this has no effect "
+                 "on the block restrictions of kernels, BCs, or other block-restrictable objects. "
```
This is confusing to me. You say in the doc-string for `block` that this is the default block restriction for kernels/materias/etc. but here you are warning users that it doesn't affect their block restriction. Which is it?

lindsayad replied:
> the expectation is that users should set `[GlobalParams]/block` directly. If we adopt this approach, he recommended that we programmatically detect whether `[GlobalParams]/block` is set, and issue a warning if it is not.

I don't agree with this approach. Having the only blessed way to set an object's parameter, in this case `Problem/block`, be from the `GlobalParams` is totally different from how users are used to interacting with parameters. `GlobalParams` are meant to be overridden by individual block specifications; individual block specifications are not meant to be discouraged in favor of `GlobalParams`. I think that `FEProblemBase` behavior should be the same whether `block` is set from the `GlobalParams` or in the `Problem` block itself and the latter should not issue a warning

> * Gary also pointed out that setting `block` in `[Problem]` alone does not enforce any restriction on block-restrictable objects like Kernels or Materials

Perhaps it doesn't currently, but what is stopping you from using the information in `Problem/block` to modify the `block` parameter of the `InputParameters` that flow through the `FEProblemBase`?

lindsayad reviewed (changes requested):
The code in the `FEProblemBase` constructor is too complex. It's becoming a spaghetti mess. I don't think there should be a default for `block`, or if we're going to have one there needs to be comments about why it's there and what an empty default means. We are multiplying the ways that a user can set block restriction. We have the standard sub-block, global params, and now this parameter. I don't want to have to have this constant conjunction of `_has_block_in_global` and `isParamSetByUser("block")`; is there a way to avoid it? If not, adding a static method or two with good doxygen for use in the constructor would be helpful.

I will continue to support this additive domain work but it is making for some head-scratching code

hugary1995 commented:
@lindsayad sorry this was confusing -- this PR is based on the other PR for default block restriction. So I guess you meant to leave the above comments there. I've addressed your comments with this commit https://github.com/idaholab/moose/pull/30300/commits/ebfec5e1a47351b98df595d86cf377780ebb66dd. The commit right after that removes those deprecated parameters from input files.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2933994/), step Clang format on d83ee36 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/30630/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/30630/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 4fe65bd151688ec8c5b363b37f3b660628d0f747`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2946451/), step Docs: sync website on 1f0790e wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/30630/site)

This comment will be updated on new commits.

GiudGiud commented on framework/include/systems/SystemBase.h:
```diff
@@ -424,6 +424,8 @@ class SystemBase : public libMesh::ParallelObject, public ConsoleStreamInterface
    */
   virtual NumericVector<Number> & serializedSolution();
 
+  virtual void cleanSerializedSolution();
```
docstring

GiudGiud commented on framework/include/userobjects/NodalPatchRecoveryBase.h:
```diff
@@ -12,6 +12,32 @@
+inline void
+hash_combine(std::size_t & seed, std::size_t value)
+{
+  seed ^= value + 0x9e3779b9 + (seed << 6) + (seed >> 2);
```
where did you get this from?

GiudGiud replied:
and does it only work for uint?

ChengHauYang replied:
Thanks for the question! I picked up that hash_combine pattern from an AI suggestion. I'm currently using it to cache a mapping from std::vector<dof_id_type> to RealEigenVector, and I would like to hear from you better suggestions if there's a more robust or standard way to hash std::vector<dof_id_type>. I would appreciate any advice!

GiudGiud replied:
@hugary1995 would you know where we can find some hashing functions for standard types?

hugary1995 replied:
I like the idea of caching query dofs, but I don't like the idea of hashing here. Can we just store the unhashed vector of dof ids as the cache lookup key? Upon every query, you need to hash the query dofs for lookup, where hashing is O(n) (based on your implementation). So if it's a linear traversal, why don't we just compare two vectors directly without any hashing? Comparing vectors is even faster because if the vector sizes don't match, it bails out without further comparison.

The only advantage of hashing that I can think of here is that we are saving some memory, i.e., a vector of size_t vs. a single size_t. But I doubt the query will ever be that large to make hashing worthy.

ChengHauYang replied:
Thanks for your feedback. I follow your suggestions to change the code in the new commit.

GiudGiud commented on framework/include/userobjects/NodalPatchRecoveryVariable.h:
```diff
@@ -0,0 +1,27 @@
+  /// Returns the variable name
+
+private:
+  const VariableValue & _var; ///< Variable value
+
+  const VariableName _var_name; ///< Variable name
```
```suggestion
  /// Variable value
  const VariableValue & _var;

   /// Variable name
  const VariableName _var_name;
```

GiudGiud commented on framework/include/userobjects/NodalPatchRecoveryVariable.h:
```diff
@@ -0,0 +1,27 @@
+protected:
+  /// Returns the recovered value at the quadrature point (_qp)
+  Real computeValue() override { return _var[_qp]; };
+  /// Returns the variable name
```
```suggestion
```

GiudGiud commented on test/tests/meshmodifiers/ic_for_newlyactivated_nodes/gold/external_moving_boundary_no_unsolved_blocks_all_IC_default_out.e-s003:
do we really get much from checking all 4 steps? could we just do the last one?

ChengHauYang replied:
Yes. I kept the last one for checking now.

GiudGiud commented on test/tests/meshmodifiers/ic_for_newlyactivated_nodes/tests:
```diff
@@ -0,0 +1,174 @@
+      type = 'Exodiff'
+      input = 'external_moving_boundary_npr.i'
+      exodiff = 'external_moving_boundary_npr_out.e external_moving_boundary_npr_out.e-s002 external_moving_boundary_npr_out.e-s003 external_moving_boundary_npr_out.e-s004'
+      detail = 'The system shall set initial conditions on newly activated nodes during the simulation.'
```
```suggestion
      detail = 'using the initial conditions defined ....'
```

details gets appended after the requirement. fix others please

ChengHauYang replied:
Thanks. I have fixed them in tests.

GiudGiud commented on test/tests/meshmodifiers/ic_for_newlyactivated_nodes/tests:
```diff
@@ -0,0 +1,174 @@
+  design = 'ElementSubdomainModifier.md'
+
+  [group]
+    requirement = 'The framework shall include the ability to set initial conditions on newly activated nodes during simulation.'
```
```suggestion
    requirement = 'The framework shall include the ability to set initial conditions on newly activated nodes during simulation'
```

GiudGiud commented on test/tests/meshmodifiers/ic_for_newlyactivated_nodes/tests:
```diff
@@ -0,0 +1,174 @@
+      type = 'Exodiff'
+      input = 'external_moving_boundary_no_unsolved_blocks_all_IC_default.i'
+      exodiff = 'external_moving_boundary_no_unsolved_blocks_all_IC_default_out.e external_moving_boundary_no_unsolved_blocks_all_IC_default_out.e-s002 external_moving_boundary_no_unsolved_blocks_all_IC_default_out.e-s003 external_moving_boundary_no_unsolved_blocks_all_IC_default_out.e-s004'
+      detail = 'The system shall run with `ic_strategy = IC_DEFAULT` without requiring `nodal_patch_recovery_uo` or `unsolved_blocks`.'
```
please dont use parameter names or values. only english in SQA doco, no moose-talk

GiudGiud commented on test/tests/meshmodifiers/ic_for_newlyactivated_nodes/tests:
```diff
@@ -0,0 +1,174 @@
+  []
+
+  [error_test]
+    requirement = 'The system shall throw errors when invalid IC strategies are used or when input vectors have inconsistent sizes.'
```
```suggestion
    requirement = 'The system shall throw an error '
```

GiudGiud commented on test/tests/meshmodifiers/ic_for_newlyactivated_nodes/tests:
```diff
@@ -0,0 +1,174 @@
+      type = RunException
+      input = 'ic_strategy_not_valid.i'
+      expect_err = 'Invalid option "IC_POLYNOMIAL_WHOLE" in MooseEnum.'
+      detail = 'The system shall throw an error if an invalid IC strategy name is provided.'
```
```suggestion
      detail = ' if an invalid initial condition strategy name is provided when modifying element subdomains'
```

GiudGiud commented on test/tests/meshmodifiers/ic_for_newlyactivated_nodes/tests:
```diff
@@ -0,0 +1,174 @@
+
+    [error_ic_strategy_not_valid]
+      type = RunException
+      input = 'ic_strategy_not_valid.i'
```
ideally you would not create an input for every error. just create a base input, and use cli_args to trigger each error. 

having invalid inputs is a bad idea documentation-wise, they show up on the object doco page

GiudGiud commented on framework/src/userobjects/NodalPatchRecoveryVariable.C:
```diff
@@ -0,0 +1,21 @@
+#include "NodalPatchRecoveryVariable.h"
```
missing copyright

GiudGiud commented on framework/src/userobjects/NodalPatchRecoveryVariable.C:
```diff
@@ -0,0 +1,21 @@
+  InputParameters params = NodalPatchRecoveryBase::validParams();
+
+  params.addRequiredCoupledVar("var", "The scalar variable to recover at nodes");
+
+  // params.addParam<unsigned int>("component", 0, "Component index (if variable has multiple)");
```
```suggestion
```

GiudGiud commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -3709,6 +3709,65 @@ FEProblemBase::projectInitialConditionOnCustomRange(ConstElemRange & elem_range,
+      DenseVector<Number> vals(var.order());
+      ic->compute(vals);
+
+      const unsigned int n_SCALAR_dofs = var.dofIndices().size();
```
```suggestion
      const unsigned int n_scalar_dofs = var.dofIndices().size();
```

see variable naming conventions

GiudGiud commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -3709,6 +3709,65 @@ FEProblemBase::projectInitialConditionOnCustomRange(ConstElemRange & elem_range,
+      const unsigned int n_SCALAR_dofs = var.dofIndices().size();
+      for (unsigned int i = 0; i < n_SCALAR_dofs; i++)
+      {
+        const dof_id_type global_index = var.dofIndices()[i];
```
```suggestion
        const auto global_index = var.dofIndices()[i];
```

GiudGiud commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -3709,6 +3709,65 @@ FEProblemBase::projectInitialConditionOnCustomRange(ConstElemRange & elem_range,
+
+  for (auto & nl : _nl)
+    nl->solution().close();
+  _aux->solution().close();
```
you do this again right below

GiudGiud replied:
dont think we have a precendence concern with scalar ics

GiudGiud commented on framework/include/loops/ComputeInitialConditionThread.h:
```diff
@@ -33,4 +36,7 @@ class ComputeInitialConditionThread
   FEProblemBase & _fe_problem;
   THREAD_ID _tid;
+
+  /// @brief the names of target variables for which the initial conditions are applied
+  const std::set<std::string> _target_var_names;
```
this is not used by the usual constructor so it s just left uninitialized. This is unexpected for anyone deriving from this class

GiudGiud commented on framework/include/userobjects/NodalPatchRecoveryVariable.h:
```diff
@@ -0,0 +1,27 @@
+#pragma once
```
missing copyright

GiudGiud replied:
still missing

GiudGiud commented on framework/include/userobjects/NodalPatchRecoveryVariable.h:
```diff
@@ -0,0 +1,27 @@
+ *
+ * Derived from NodalPatchRecoveryBase.
+ * Reads any coupled variable using `coupledValue(<var_name>)[_qp]`.
+ * Suitable for all scalar variables; for vector variables, split them in the input file as <var>_x,
```
```suggestion
 * Suitable for all field scalar-valued variables; for vector variables, split them in the input file as <var>_x,
```

GiudGiud reviewed (changes requested):
havent looked at the two main files yet

hugary1995 commented on framework/include/loops/ComputeBoundaryInitialConditionThread.h:
```diff
@@ -27,7 +27,15 @@ class ComputeBoundaryInitialConditionThread
 
+  // Set IC on specific variable names
+  ComputeBoundaryInitialConditionThread(FEProblemBase & fe_problem,
+                                        const std::set<std::string> & target_var_names);
```
we use VariableName instead of string in moose.

hugary1995 commented on framework/include/meshmodifiers/ElementSubdomainModifierBase.h:
```diff
@@ -12,6 +12,16 @@
+#include "NodalPatchRecoveryBase.h"
+#include "KDTree.h"
+
+enum class ICStrategy
```
Remove the IC_ prefix in the enum

hugary1995 replied:
What's the difference between IC_POLYNOMIAL and IC_POLYNOMIAL_WHOLE_SOLVED_DOMAIN?

hugary1995 commented on framework/include/meshmodifiers/ElementSubdomainModifierBase.h:
```diff
@@ -141,8 +163,80 @@ class ElementSubdomainModifierBase : public ElementUserObject
+  /// Range of reinitialized nodes
+  std::unique_ptr<ConstNodeRange> _reinitialized_node_range;
+  /// Range of reinitialized nodes on the displaced mesh
+  std::unique_ptr<ConstNodeRange> _reinitialized_displaced_node_range;
```
why do we need a separate range for displaced nodes?

ChengHauYang replied:
I removed _reinitialized_displaced_node_range

hugary1995 commented on framework/include/meshmodifiers/ElementSubdomainModifierBase.h:
```diff
@@ -141,8 +163,80 @@ class ElementSubdomainModifierBase : public ElementUserObject
+  /// Range of reinitialized nodes extracted from boundary nodes (displaced mesh)
+  std::unique_ptr<ConstNodeRange> _reinitialized_displaced_node_range_from_bnd_nodes;
+
+  /// The strategy used to apply IC on newly activated nodes
```
correct the doc string

hugary1995 commented on framework/include/meshmodifiers/ElementSubdomainModifierBase.h:
```diff
@@ -141,8 +163,80 @@ class ElementSubdomainModifierBase : public ElementUserObject
+  std::vector<unsigned int> _ic_vars_number;
+
+  /// @brief map from variable number to the index of the nodal patch recovery user object in `_npr_vec`
+  std::unordered_map<unsigned int, unsigned int> _var_number2_npr_idx;
```
change 2 to to

ChengHauYang replied:
Fixed

hugary1995 commented on framework/include/meshmodifiers/ElementSubdomainModifierBase.h:
```diff
@@ -141,8 +163,80 @@ class ElementSubdomainModifierBase : public ElementUserObject
+
+  /// Global collection of all newly activated node IDs, gathered across all processors.
+  /// This set is independent of processor ownership and ensures consistency in parallel runs.
+  std::vector<dof_id_type> _complete_reinitialized_nodes;
```
what's the usage of this member?

ChengHauYang replied:
I removed this part. It was used to gather the complete set of globally reinitialized nodes from all reinitialized elements, ensuring that every processor holds the same copy. Later, we filter these globally reinitialized nodes and retain only those owned by the current processor by checking whether node.processor_id() equal to _mesh.processor_id().

hugary1995 commented on framework/include/meshmodifiers/ElementSubdomainModifierBase.h:
```diff
@@ -141,8 +163,80 @@ class ElementSubdomainModifierBase : public ElementUserObject
+
+  /// IC_POLYNOMIAL_THRESHOLD related parameters
+  /// @brief Threshold for checking the closeness of element numbers in polynomial extrapolation
+  int _nearby_element_threshold = 1;
```
what's "closeness"?

ChengHauYang replied:
I fixed this in the latest commit with clearer comments:
Basically, _nearby_element_threshold refers to the minimum number of nearby elements required in the polynomial extrapolation patch.

hugary1995 commented on framework/include/meshmodifiers/ElementSubdomainModifierBase.h:
```diff
@@ -141,8 +163,80 @@ class ElementSubdomainModifierBase : public ElementUserObject
+  std::vector<dof_id_type> _kd_tree_sequence_elem_id_map;
+
+  /// @brief Minimum diagonal length (criteria) for the k-d tree search
+  double _min_diag_length = std::numeric_limits<double>::max();
```
what does diagonal mean?

ChengHauYang replied:
I’ve added a clearer comment in the latest commit.
In short, _min_diag_length is the smallest diagonal of the solved elements’ loose bounding boxes, and the k-d tree search radius is _nearby_element_threshold * _min_diag_length.

hugary1995 commented on framework/include/meshmodifiers/ElementSubdomainModifierBase.h:
```diff
@@ -141,8 +163,80 @@ class ElementSubdomainModifierBase : public ElementUserObject
+  double _radius_search_threshold = -1;
+
+  /// Number of nonlinear variables in the system
+  const int _number_of_nl_variables;
```
why do we need this member?

hugary1995 commented on framework/src/meshmodifiers/ElementSubdomainModifierBase.C:
```diff
@@ -69,6 +69,12 @@ ElementSubdomainModifierBase::validParams()
       "(IDs or names) is provided, then only elements whose new subdomain is in the list will be "
       "reinitialized. If an empty list is set, then no elements will be reinitialized.");
+
+  params.addParam<std::vector<SubdomainID>>(
+      "reinitialize_subdomain_ids",
+      {},
+      "List of subdomain IDs to reinitialize. If empty, all subdomains will be reinitialized.");
```
Why do we need this parameter if we already have `reinitialize_subdomains`?

ChengHauYang replied:
Yes. I removed this parameter. Thanks!

hugary1995 commented on framework/include/problems/FEProblemBase.h:
```diff
@@ -891,6 +891,39 @@ class FEProblemBase : public SubProblem, public Restartable
+      ConstBndNodeRange & bnd_node_range,
+      const std::set<VariableName> & selected_var_names);
+
+  /**
+   * Project an initial condition given by a polynomial onto selected elements and nodes
+   * for a set of variables.
+   *
+   * Coefficient order depends on the problem dimension:
+   *
+   * - 1D (x):         [c], [c, x], [c, x, x^2]
+   * - 2D (x, y):      [c], [c, y, x], [c, y, x, y^2, xy, x^2]
+   * - 3D (x, y, z):   [c], [c, z, y, x], [c, z, y, x, z^2, zy, zx, y^2, yx, x^2]
+   *
+   * Terms not included are assumed zero.
+   *
+   * \param elem_range       Element range to project on (non-nodal)
+   * \param bnd_nodes        Boundary nodes to include (non-nodal)
+   * \param node_range       Node range for nodal variables
+   * \param coef             Polynomial coefficients
+   * \param target_var_names Set of variable names to project
+   */
+  void projectFunctionOnCustomRangeForSpecificVars(ConstElemRange & elem_range,
+                                                   ConstNodeRange & bnd_nodes,
+                                                   ConstNodeRange & node_range,
+                                                   const RealEigenVector & coef,
+                                                   const std::set<VariableName> & target_var_names);
```
This implementation is too specific to live inside FEProblemBase. Can't we replace the `RealEigenVector` with an actual functor/lambda?

ChengHauYang replied:
I implemented a more general one that takes functions (poly_func, poly_func_grad) so it can work with any user-defined function.

hugary1995 commented on framework/include/problems/FEProblemBase.h:
```diff
@@ -891,6 +891,39 @@ class FEProblemBase : public SubProblem, public Restartable
+   * Project initial conditions onto a specified set of elements and boundary nodes
+   * for a user-defined subset of variables.
+   */
+  void projectInitialConditionOnCustomRangeForSpecificVars(
```
I'd prefer to overload the original method rather than create a new method with a different name.

ChengHauYang replied:
I’ve combined the two functions into a single projectInitialConditionOnCustomRange() with default arguments for target_var_names and TargetVarUsageForIC.

hugary1995 commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -3709,6 +3709,228 @@ FEProblemBase::projectInitialConditionOnCustomRange(ConstElemRange & elem_range,
   _aux->solution().localize(*_aux->sys().current_local_solution, _aux->dofMap().get_send_list());
 }
 
+void
+FEProblemBase::projectInitialConditionOnCustomRangeForSpecificVars(
+    ConstElemRange & elem_range,
+    ConstBndNodeRange & bnd_nodes,
+    const std::set<VariableName> & target_var_names)
+{
```
It seems that there's quite some code duplication with the other method.

ChengHauYang replied:
I’ve combined the two functions into a single projectInitialConditionOnCustomRange().

hugary1995 commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -3709,6 +3709,228 @@ FEProblemBase::projectInitialConditionOnCustomRange(ConstElemRange & elem_range,
+    es.parameters.set<Real>("coef_xx") = get(9);
+  }
+
+  // Define projection function
+  auto poly_func = [](const Point & p,
+                      const Parameters & parameters,
+                      const std::string &,
+                      const std::string &) -> Number
+  {
```
Would be nice if we can define the lambda outside and pass it in as an argument.

ChengHauYang replied:
Yes. Right now, the function would be defined outside this member function inside FEProblemBase.

ChengHauYang commented on framework/src/loops/ComputeBoundaryInitialConditionThread.C:
```diff
@@ -17,13 +17,22 @@
+ComputeBoundaryInitialConditionThread::ComputeBoundaryInitialConditionThread(
+    FEProblemBase & fe_problem, const std::set<VariableName> & target_var_names)
+  : ThreadedNodeLoop<ConstBndNodeRange, ConstBndNodeRange::const_iterator>(fe_problem),
+    _target_var_names(target_var_names)
```
I addressed this by adding a TargetVarUsageForIC argument to the constructor, which allows selecting between ONLY_LIST and SKIP_LIST for the target_var_names.

The updated behavior is as follows:
- If we pass an empty set with SKIP_LIST (the default) and leave target_var_names = {} (default), initial conditions are applied to all variables.
- If we pass a specific set with ONLY_LIST, initial conditions are applied only to the variables in the set. This is what I use in the NPR case to set ICs for particular variables.

ChengHauYang commented on framework/src/loops/ComputeInitialConditionThread.C:
```diff
@@ -14,13 +14,19 @@
+
+ComputeInitialConditionThread::ComputeInitialConditionThread(
+    FEProblemBase & fe_problem, const std::set<VariableName> & target_var_names)
+  : _fe_problem(fe_problem), _target_var_names(target_var_names)
```
I addressed this using TargetVarUsageForIC (ONLY_LIST or SKIP_LIST). See comments above.

hugary1995 committed:
remove variable name in NodalPatchRecoveryBase

hugary1995 committed:
minor clean up in ElementSubdomainModifierBase.C

hugary1995 committed:
done with ElementSubdomainModifierBase

hugary1995 committed:
reorganize NodalPatchRecoveryBase; remove use_specific_elements

hugary1995 committed:
renamed test folder; discovered a parallel bug

ChengHauYang committed:
Fix parallel consistency in ESM reinit: gather global reinitialized elements for correct neighbor patch and DOF restoration; refactor sync interface and update tests

ChengHauYang committed:
Refactor semi-local reinitialization for ghosted elements in patch recovery

- Replaced `_global_reinitialized_elems` with `_semi_local_reinitialized_elems` to include ghosted and local elements without global communication.
- Rewrote logic in `findReinitializedElemsAndNodes()` to push element IDs to neighbor processors via `push_parallel_vector_data`.
- Updated all relevant loops (e.g., `storeOverriddenDofValues`, `gatherPatchElements`) to use `_semi_local_reinitialized_elems`.

ChengHauYang committed:
[WIP] dofs communicated by libmesh testing

hugary1995 committed:
Fix parallel projection bug

hugary1995 committed:
remove redundant checks

hugary1995 commented:
@GiudGiud @lindsayad or @roystgnr can we get a review on the framework level changes?

The main beef is in `ElementSubdomainModifierBase`, and I've already reviewed that part fairly thoroughly. The rest of the changes in framework are pretty light.

hugary1995 commented:
The PR has adapted quite a lot since its creation, so I took the liberty of rewriting the PR description. @ChengHauYang see if what I wrote makes sense and reflects what you have now.

ChengHauYang committed:
minor cleanup

ChengHauYang committed:
[Fix] Ensure consistent results for IC using `POLYNOMIAL_NEARBY`
`Sort` patch_elems before std::unique to remove duplicates, ensuring consistent results for every run.

hugary1995 commented on framework/src/meshmodifiers/ElementSubdomainModifierBase.C:
```diff
@@ -1056,6 +1056,8 @@ ElementSubdomainModifierBase::gatherPatchElements(const VariableName & var_name,
   _mesh.comm().allgather(patch_elems);
 
+  // Remove duplicates from the patch elements (espcially important for POLYNOMIAL_NEARBY)
+  std::sort(patch_elems.begin(), patch_elems.end());
```
Why do we need to sort here?

ChengHauYang committed:
[minor] fix Markdown

ChengHauYang committed:
fix(distributed-mesh): make NPR/ESM robust in true distributed runs

Without `--distributed-mesh`, element metadata is globally accessible even if partitioned, but with `--distributed-mesh` remote elements become nullptr on non-owner ranks, causing segfaults when accessing `centroids()` or `processor_id()`.

This commit updates patch-recovery and nearby-search logic to work correctly in true distributed mode by gathering (centroid, elem_id) pairs across ranks when distributed and only elem_id in replicated mode, building the KD-tree from the gathered points and mapping query results back to global elem_ids.

In NodalPatchRecoveryBase, cache keys are sorted and deduplicated before comparison, and null checks are introduced for remote elements. The gatherSendList() method is reworked so that in distributed mode it uses Parallel::push_parallel_vector_data to pass (owner_pid, elem_id) to the other ranks (except for the current rank), while in replicated mode it continues to allow direct cross-rank access.

ChengHauYang committed:
Clarify optional `specific_elems` behavior in NodalPatchRecoveryBase and improve IC thread control.

- Updated docstring for `sync()` and `gatherSendList()` to explain default all-semi-local-element synchronization vs. targeted subset mode for algorithmic needs.
- Restricted targeted IC application logic by making specific-variable constructors private and accessible only to FEProblemBase (private + friend pattern) for both CIC and CIBC threads.
- Replaced deprecated `resetSerializedSolution()` with direct `zero_clone()` vector creation.
- Removed unused includes, parameters, and unnecessary diffs.

ChengHauYang commented on framework/include/userobjects/NodalPatchRecoveryBase.h:
```diff
@@ -29,10 +29,32 @@ class NodalPatchRecoveryBase : public ElementUserObject
+  /**
+   * @brief Synchronizes local matrices and vectors (_Ae, _be) across processors
+   *
+   * Gathers and distributes data required for evaluable elements (possibly from other processors)
```
Thanks for the comment -- specific_elems is optional and is intended for cases where the monomial basis needs to be built from a specific patch, or even a subset of that patch.

If not provided, the method defaults to synchronizing all ghosted evaluable elements, so there is no risk of missing data in the general case. When a subset is provided, it is an intentional choice for performance or algorithmic reasons, and the caller is responsible for specifying the needed elements.

I’ve updated the docstring to clarify this behavior:

```
/**
 * @brief Synchronizes local matrices and vectors (_Ae, _be) across processors.
 *
 * If @p specific_elems is provided, only those elements will be synchronized.
 * This is typically used when the monomial basis needs to be built from a
 * specific patch, or even a subset of that patch.
 *
 * If @p specific_elems is not provided, all ghosted evaluable elements are
 * synchronized (default behavior).
 */
```

hugary1995 replied:
Right now this is only called by element subdomain modifiers when retrieving polynomial coefficients for extrapolation and projection. Everything happens in the framework. @ChengHauYang I think @lindsayad is hinting at making this a private/protected method and friend the ESM. You can also separate this method into two overloads, i.e.
```
void sync(std::vector<dof_id_type> & specific_elems);
void sync();
```
The first overload operates on specific (possibly nonlocal) elements. The second overload operates on ghosted evaluable elements. Both overloads can be private, and we only need to friend the first overload to ESM.

lindsayad replied:
I would very much like to lock this into the framework yes

hugary1995 replied:
How do you want to implement the "lock". We now mark the methods as private and explicitly friend the ElementSubdomainModifierBase, which is effectively "locking" the access I suppose. Do you want a passkey instead?

ChengHauYang commented on framework/include/userobjects/NodalPatchRecoveryBase.h:
```diff
@@ -41,6 +63,11 @@ class NodalPatchRecoveryBase : public ElementUserObject
   unsigned int _qp;
 
 private:
+  /// Iterates over all evaluable elements and records their IDs in a query map
```
Thanks!

The default case (no `specific_elems`) synchronizes **all semi-local evaluable elements**  
(including ghost elements) across processors.  

`specific_elems` is optional and only used when we intentionally want to  
synchronize a targeted subset, e.g., building a monomial basis from a specific patch or part of it. 

I updated the docstring for clarification.

hugary1995 replied:
Same here, @ChengHauYang let's keep this as unresolved until we reach a consensus with the reviewer.

ChengHauYang commented on framework/src/loops/ComputeBoundaryInitialConditionThread.C:
```diff
@@ -44,8 +46,16 @@ ComputeBoundaryInitialConditionThread::onNode(ConstBndNodeRange::const_iterator
     const auto & ics = warehouse.getActiveBoundaryObjects(boundary_id, _tid);
     for (const auto & ic : ics)
+    {
+
```
Fixed

ChengHauYang commented on framework/include/loops/ComputeBoundaryInitialConditionThread.h:
```diff
@@ -21,13 +22,19 @@ class ComputeBoundaryInitialConditionThread
+  // Set IC on specific variables
+  ComputeBoundaryInitialConditionThread(
+      FEProblemBase & fe_problem,
+      const std::optional<std::set<VariableName>> & target_vars = std::nullopt);
```
Thanks -- agreed on keeping tight control within the framework. I fixed that for both CIBC and CIC.

I implemented the **private + friend** approach.

ChengHauYang commented on framework/include/loops/ComputeInitialConditionThread.h:
```diff
@@ -20,7 +21,11 @@ class FEProblemBase;
 public:
-  ComputeInitialConditionThread(FEProblemBase & fe_problem);
+  // Set IC on specific variables
+  ComputeInitialConditionThread(
```
Fixed. Thanks!

ChengHauYang commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -7917,7 +7974,9 @@ FEProblemBase::initialAdaptMesh()
-            /*intermediate_change=*/false, /*contract_mesh=*/true, /*clean_refinement_flags=*/true);
+            /*intermediate_change=*/false,
+            /*contract_mesh=*/true,
+            /*clean_refinement_flags=*/true);
```
Yes. I tried to remove these unnecessary diffs

ChengHauYang committed:
[Minor fix] Initialize _target_vars to avoid undefined access. Uninitialized _target_vars_ could cause unexpected behavior

hugary1995 commented on framework/src/systems/SystemBase.C:
```diff
@@ -1,5 +1,5 @@
 //* This file is part of the MOOSE framework
-//* https://mooseframework.inl.gov
+//* https://www.mooseframework.org
```
I think the correct version is .inl.gov. @ChengHauYang can you double check?

hugary1995 commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -3735,6 +3742,55 @@ FEProblemBase::projectInitialConditionOnCustomRange(ConstElemRange & elem_range,
+  temp_vec.close();
+
+  // Get the dof indices to copy
+  DofMap & dof_map = sys.dofMap();
+  std::set<dof_id_type> dof_indices;
+  std::vector<dof_id_type> elem_dof_indices;
+
+  for (const auto & elem : elem_range)
+  {
+    dof_map.dof_indices(elem, elem_dof_indices, var_num);
+    dof_indices.insert(elem_dof_indices.begin(), elem_dof_indices.end());
+  }
+  std::vector<dof_id_type> dof_indices_v(dof_indices.begin(), dof_indices.end());
```
@ChengHauYang let's leave non-trivial comments to the reviewer to mark as resolved. That way we can make sure the reviewer agrees with our explanation.

hugary1995 commented on framework/include/userobjects/NodalPatchRecoveryBase.h:
```diff
@@ -29,10 +29,32 @@ class NodalPatchRecoveryBase : public ElementUserObject
+   * @param elem_ids    Ids of the elements in the patch
+   * @return The coefficients of the polynomial
+   *
+   * @warning This method shall not be called within a threaded region as it modifies some mutable
```
Yeah, I think it's actually safe to use this even in a threaded region for the reason @lindsayad mentioned. We can remove the warning and mark this method as non-const.

ChengHauYang committed:
[modify(nodal-patch-recovery)] split coefficient computation into const and non-const versions, add duplicate removal helper
Introduced getCoefficientsNoCache() as a const method that always recomputes polynomial coefficients without caching.
Updated nodalPatchRecovery() to use getCoefficientsNoCache() for const contexts.
Modified getCoefficients() to be non-const, check cache before computing, and delegate recomputation to getCoefficientsNoCache().

ChengHauYang committed:
[modify ESM] add timestep-based restore of moved elements for restep handling
Introduced timestepSetup() to detect restep conditions by comparing _t_step with a cached _t_step_old. If a restep is detected, previously moved elements are restored to their original subdomains before proceeding.

ChengHauYang committed:
[Fix] Robust subdomain ID restore on timestep repeats and rejected solves

- Track previous step completion with `_has_done_previous_step`.
- In `timestepSetup()`, if the timestep did **not** advance but the previous step completed, automatically restore subdomain IDs (handles restep/rejected solves).
- If called multiple times in one step without advancing the timestep and `_has_done_previous_step` is false, do **not** revert subdomain IDs (supports workflows like XFEM).
- `_has_done_previous_step` is set true after a successful step and subdomain change, and reset to false at each timestep.

Prevents mesh inconsistency when re-trying steps or intentionally reusing the modifier within a single timestep.

ChengHauYang committed:
Revert commit | `skip_restore_subdomain_changes` in ESM, which prevents reverting the subdomain changes in XFEM

hugary1995 commented:
The documentation failure is because NodalPatchRecoveryVariable is missing class description. You probably forgot to do that in the validParams method.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/3139051/), step Generate coverage on d77c6b3 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>9fc4b0</th>
      <th colspan=4>#30630 d77c6b</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.59%</td>
      <td>85.60%</td>
      <td>+0.01%</td>
      <td>87.72%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>116490</td>
      <td>116850</td>
      <td>+360</td>
      <td>407</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>19613</td>
      <td>19651</td>
      <td>+38</td>
      <td>57</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/30630/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/30630/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/30630/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/30630/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/30630/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/30630/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/30630/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/30630/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/30630/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/30630/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/30630/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/30630/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/30630/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/30630/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/30630/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/30630/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/30630/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/30630/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/30630/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/30630/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/30630/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/30630/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/30630/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/30630/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/30630/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/30630/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/30630/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/30630/coverage/stochastic_tools)
- [`subchannel`](https://mooseframework.inl.gov/docs/PRs/30630/coverage/subchannel)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/30630/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/30630/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 87.72% is less than the suggested 90.0%



This comment will be updated on new commits.

ChengHauYang committed:
[Patch Recovery] Switch to NodalPatchRecoveryVariable for variable-specific patch recovery

- Use NodalPatchRecoveryVariable instead of the base class in ESMBase.
- Add class description to fix documentation errors.
- Assert variable name matches during setup.

ChengHauYang committed:
turn on parallel and distributed testing for mortar/mesh_modification/mortar_moving_subdomains

ChengHauYang committed:
remove unnecessary diffs

hugary1995 committed:
Minor change in ESMBase applyMovingBoundaryChanges

hugary1995 committed:
Revert "turn on parallel and distributed testing for mortar/mesh_modification/mortar_moving_subdomains"

This reverts commit 62f5ce1e85a2fbc93168b641d2a2d3b0c71da35a.

lindsayad committed:
Reapply "turn on parallel and distributed testing for mortar/mesh_modification/mortar_moving_subdomains"

This reverts commit 68f1d6314870acbb456b079dff45e83a582e29c3.

lindsayad committed:
Make displaced and mesh mod test LBB stable

hugary1995 committed:
Apply boundary conditions on the correct moving boundaries

ChengHauYang committed:
Refactor NodalPatchRecoveryBase sync logic

- Split sync into two private overloads: sync() for all ghosted elements, and sync(const std::vector<dof_id_type>&) for specific elements. Made ElementSubdomainModifierBase a friend class, ensuring these methods are not exposed to the user.

hugary1995 committed:
remove extra ;

GiudGiud commented on framework/doc/content/source/userobjects/NodalPatchRecoveryVariable.md:
```diff
@@ -0,0 +1,11 @@
+# NodalPatchRecoveryVariable
```
Thoughts about making this a Corrector? It's another system without a base class which is iffy, but in the future there might be some functionality we can share

hugary1995 replied:
Yeah, the patch recovery capability is pretty standalone in the sense that it is responsible for projecting some field onto arbitrary global function space, and that task is fairly generic.

It doesn't prevent you from implementing a patch-recovery-based corrector though. But that should probably be a separate class that calls into this one.

Now that we are on this subject, as a related note, I think we would benefit from moving the patch recovery logic into the framework itself (e.g., integrating it into SubProblem or perhaps SystemBase). I'm probably guilty of implementing too many capabilities as userobjects... I'll think more about this, and please let me know what your opinions are.

lindsayad replied:
I'm always all for upstreaming things

GiudGiud commented on framework/include/meshmodifiers/ElementSubdomainModifierBase.h:
```diff
@@ -95,16 +130,43 @@ class ElementSubdomainModifierBase : public ElementUserObject
+  /// Whether this is a re-step
+  bool _restep;
+
+  /// Skipping restroring the subdomain changes if the timestep is not advanced
```
```suggestion
  /// Skipping restoring the subdomain changes if the timestep is not advanced
```

ChengHauYang replied:
Thanks for carefully checking! Fixed!

GiudGiud commented on framework/include/meshmodifiers/ElementSubdomainModifierBase.h:
```diff
@@ -136,13 +198,63 @@ class ElementSubdomainModifierBase : public ElementUserObject
+  std::map<VariableName, std::pair<std::vector<dof_id_type>, std::vector<Number>>>
+      _overridden_values_on_reinit_elems;
+
+  /// The strategy used to apply IC on newly activated nodes
```
```suggestion
  /// The strategy used to apply IC on newly activated elements
```
nodes? does this only work for first lagrange?

ChengHauYang replied:
It is a function space independent implementation. I just fixed this.

GiudGiud commented on framework/include/userobjects/NodalPatchRecoveryBase.h:
```diff
@@ -29,10 +31,30 @@ class NodalPatchRecoveryBase : public ElementUserObject
+  const RealEigenVector getCoefficientsNoCache(const std::vector<dof_id_type> & elem_ids) const;
+
+  /**
+   * Non-const version of getCoefficients. This will recompute the coefficients
+   * even if they are already cached.
```
so which one is the one that does not recompute the coefficients?

GiudGiud replied:
ok this is not right?
seems the cached is used

ChengHauYang replied:
We fixed the comments with

```
  /**
   * Compute coefficients, using cached values if available, and store any newly computed
   * coefficients in cache
   * The coefficients returned by this function are intended to be multiplied with the multi-index
   * table to build the corresponding polynomial. For details on the multi-index table, refer to the
   * comments in the multiIndex() function.
   *
   * @param elem_ids Ids of the elements in the patch
   * @return The coefficients of the polynomial
   */
```

Thanks!

GiudGiud commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -3735,6 +3757,53 @@ FEProblemBase::projectInitialConditionOnCustomRange(ConstElemRange & elem_range,
+                                            const libMesh::Parameters & params,
+                                            const VariableName & target_var)
+{
+  const auto & var = getStandardVariable(0, target_var);
```
assert that you are not in a threaded region

GiudGiud reviewed (commented):
I ll finish later today or next week

bwspenc commented on framework/src/userobjects/NodalPatchRecoveryBase.C:
```diff
@@ -130,26 +192,93 @@ NodalPatchRecoveryBase::finalize()
+    else
+      // For non-distributed meshes, element information is always accessible (not nullptr and
+      // processor ID can be retrieved), even if the element does not belong to the current
+      // processor.
```
This comment doesn't completely make sense. Please make sure it's correct and concise.

ChengHauYang replied:
Sure. I will change the comments to be

```
// Add to query_ids if this element's data is on a different processor
```

ChengHauYang commented on framework/src/meshmodifiers/ElementSubdomainModifierBase.C:
```diff
@@ -623,21 +927,50 @@ ElementSubdomainModifierBase::reinitializedBndNodeRange(bool displaced)
+  for (auto node_id : _reinitialized_nodes)
+    nodes.push_back(_mesh.nodePtr(node_id)); // displaced mesh shares the same node object
+
+  // Make some fake node iterators defining this vector of nodes
```
Sorry for the misleading. I previously had a legacy implementation that made the boundary nodes the nodes. We do not do this anymore. And, I still leave the legacy comments over there. I will remove this.

ChengHauYang commented on framework/src/userobjects/NodalPatchRecoveryBase.C:
```diff
@@ -97,6 +152,12 @@ NodalPatchRecoveryBase::initialize()
+  // make sure to clear the cached coefficients
+  // so that the next time we call nodalPatchRecovery, we will recompute the coefficients to make
+  // sure _Ae and _be has already been different but we do not use the same coefficients for the
+  // same element
```
Sure.

I will change it to be

```
  // Clear cached data to ensure coefficients are correctly recomputed in the next patch recovery
  // iteration. _Ae and _be must also be reset, as the associated patch elements may differ in the
  // upcoming iteration.
```

ChengHauYang commented on framework/include/userobjects/NodalPatchRecoveryBase.h:
```diff
@@ -29,10 +31,30 @@ class NodalPatchRecoveryBase : public ElementUserObject
-  virtual void threadJoin(const UserObject &) override;
-  virtual void finalize() override;
+  /**
+   * Const version of getCoefficients. This will always recompute the coefficients.
```
Thanks for the suggestions. I am going to change the comments to be:

```
 /**
 * Compute coefficients without reading or writing cached values, which allows this method to be
 * const
 * The coefficients returned by this function are intended to be multiplied with the multi-index
 * table to build the corresponding polynomial. For details on the multi-index table, refer to the
 * comments in the multiIndex() function.
 *
 * @param elem_ids Ids of the elements in the patch
 * @return The coefficients of the polynomial
 */
  const RealEigenVector getCoefficientsNoCache(const std::vector<dof_id_type> & elem_ids) const;

 /**
 * Compute coefficients, using cached values if available, and store any newly computed
 * coefficients in cache
 * The coefficients returned by this function are intended to be multiplied with the multi-index
 * table to build the corresponding polynomial. For details on the multi-index table, refer to the
 * comments in the multiIndex() function.
 *
 * @param elem_ids Ids of the elements in the patch
 * @return The coefficients of the polynomial
 */
  const RealEigenVector getCoefficients(const std::vector<dof_id_type> & elem_ids);
```

bwspenc replied:
I like Gary's suggestion for the names, but I'm fine how they are too.

ChengHauYang committed:
[Docs/Comments] clarify patch recovery, reinit strategies, and IC handling

- Fixed documentation in `ElementSubdomainModifier.md` and improved `NodalPatchRecoveryVariable.md`:
- Enhanced code comments in headers:
  * `///` Doxygen style in `C[B]IC.h`.
  * Improved explanations of coefficient caching, multi-index tables, and inter-processor sync in `NodalPatchRecoveryBase.h`.
  * Renamed `deduplicate` to `removeDuplicateEntries` with clearer comments.

ChengHauYang commented on framework/doc/content/source/meshmodifiers/ElementSubdomainModifier.md:
```diff
@@ -96,4 +96,56 @@ Reinitialization can be further restricted by setting the parameter `old_subdoma
+
+There are two strategies for (re)initializing variables on the *updated active* domain. The first, arguably the simplest, strategy is to project the initial condition of the variable onto the *updated active* domain.
+
+However, directly imposing such initial conditions often poses significant challenges for the solution procedure. For nonlinear material models, suboptimal initial guess could hinder convergence of the material model and/or that of the global solve. For evaluations on the deformed mesh, e.g., when geometric nonlinearity or mechanical contact is enabled, a bad initial condition of displacement cou ...
```
Fixed. Thank you!

ChengHauYang commented on framework/doc/content/source/meshmodifiers/ElementSubdomainModifier.md:
```diff
@@ -96,4 +96,56 @@ Reinitialization can be further restricted by setting the parameter `old_subdoma
+
+The second strategy aims to address these challenges. The key idea is to initialize the solution on the *updated active* domain leveraging the solution information from the *stationary active* domain. In the current implementation, we adopt the Zienkiewicz–Zhu patch recovery technique to initialize the solution field for the DoFs on *updated active* elements.
+
+The initialization algorithm starts with constructing a patch of elements from the *stationary active* domain. By definition, solution to the variable of interest already exists on the patch. The solution is first interpolated onto the quadrature points on the *stationary active* domain, and a polynomial is fitted against the interpolated variable values using a least-squares fit. A complete set  ...
```
Fixed

ChengHauYang commented on framework/doc/content/source/meshmodifiers/ElementSubdomainModifier.md:
```diff
@@ -96,4 +96,56 @@ Reinitialization can be further restricted by setting the parameter `old_subdoma
\ No newline at end of file
+!media large_media/mesh_modifiers/element_subdomain_modifier/from_into.png style=float:center;width:100%; caption=Reinitialization of only the elements which change subdomain ID from 3, to subdomain IDs 1 or 2
+
+## Variable (re)initialization on the **updated active** domain
```
Fixed!

ChengHauYang commented on framework/include/meshmodifiers/ElementSubdomainModifierBase.h:
```diff
@@ -136,13 +198,63 @@ class ElementSubdomainModifierBase : public ElementUserObject
-  /// Range of reinitialized elements on the displaced mesh
-  std::unique_ptr<ConstElemRange> _reinitialized_displaced_elem_range;
+
+  /// Semi-local Reinitialized elements:
```
I fixed this to be

```
/// Semi-local reinitialized elements: ghosted and local reinitialized elements
```

ChengHauYang commented on framework/include/userobjects/NodalPatchRecoveryBase.h:
```diff
@@ -41,6 +63,18 @@ class NodalPatchRecoveryBase : public ElementUserObject
 private:
+  /// Builds a query map of element IDs that require data from other processors.
+  /// @param specific_elems Only iterates over the provided elements and records
+  /// those belonging to a different processor.
```
Fixed

ChengHauYang commented on framework/include/userobjects/NodalPatchRecoveryBase.h:
```diff
@@ -41,6 +63,18 @@ class NodalPatchRecoveryBase : public ElementUserObject
+  /// Builds a query map of element IDs that require data from other processors.
+  /// Iterates over all semi-local evaluable elements (including ghost elements) and records those
+  /// belonging to a different processor.
+  /// Ensures that all semi-local evaluable elements are properly synchronized.
```
Fixed. Thanks!

hugary1995 commented on framework/include/loops/ComputeInitialConditionThread.h:
```diff
@@ -33,4 +35,15 @@ class ComputeInitialConditionThread
+  /// @brief the names of target variables for which the initial conditions are applied
+  const std::set<VariableName> * _target_vars = nullptr;
+
+  // Set IC on specific variables
```
I mean this file predates us. We could add doxygen documentation for all members, but then we are risking putting wrong/inaccurate descriptions... FWIW, the new constructor and member do have doxygen documentation :)

bwspenc replied:
I was just saying that the new descriptions you added should have been doxygen-style rather than `//`. Just change them where applicable to `///`. There's another one you added on line 23 that should also be `///`, and I guess the one on line 47 should really be a `//`.

hugary1995 replied:
Ah, yeah, I agree.

bwspenc commented on framework/doc/content/source/userobjects/NodalPatchRecoveryVariable.md:
```diff
@@ -0,0 +1,13 @@
+# NodalPatchRecoveryVariable
+
+This user object performs the calculations and parallel communication required to carry out the [Zienkiewicz–Zhu patch recovery](https://doi.org/10.1002/nme.1620330702) for a scalar component of a provided variable.
```
Thanks for adding the reference! However, please instead use the [!cite] syntax to reference the paper, and then add an entry for it in one of the .bib files.

bwspenc commented on framework/doc/content/source/userobjects/NodalPatchRecoveryVariable.md:
```diff
@@ -0,0 +1,11 @@
+
+This user object performs the calculations and parallel communication to perform
+the Zienkiewicz-Zhu patch recovery for a scalar component of a provided
+variable.
```
Thanks for adding the reference! However, please instead use the [!cite] syntax to reference the paper, and then add an entry for it in one of the .bib files.

ChengHauYang replied:
Sure! Thanks, @bwspenc, for your feedback!

hugary1995 commented on framework/doc/content/source/userobjects/NodalPatchRecoveryVariable.md:
```diff
@@ -0,0 +1,13 @@
+
+This user object performs the calculations and parallel communication required to carry out the [Zienkiewicz–Zhu patch recovery](https://doi.org/10.1002/nme.1620330702) for a scalar component of a provided variable.
+
+For recovering a scalar component of a material property, see the related user object: [NodalPatchRecoveryMaterialProperty](NodalPatchRecoveryMaterialProperty.md).
```
```suggestion
```

Don't really need to mention this here.

hugary1995 replied:
Whoops. I would insist on removing this though, because in the future we may add more similar userobjects, e.g., `NodalPatchRecoveryArrayVariable`, `NodalPatchRecoveryVectorVariable`, `NodalPatchRecoveryRankTwoMaterialProperty`, etc. and we don't want to keep expanding the list here -- that would be a maintenance burden.

The syntax list already does this job in my opinion. When users go to the syntax page and search for nodalpatchrecovery, they will immediately see all the relevant objects.

ChengHauYang replied:
Thanks! I have removed that in the newest commit.

hugary1995 commented on framework/doc/content/source/userobjects/NodalPatchRecoveryVariable.md:
```diff
@@ -0,0 +1,13 @@
+# NodalPatchRecoveryVariable
+
+This user object performs the calculations and parallel communication required to carry out the [Zienkiewicz–Zhu patch recovery](https://doi.org/10.1002/nme.1620330702) for a scalar component of a provided variable.
```
This isn't how we add references. Add a bibtex entry to `framework/doc/content/bib/moose.bib` and reference it like this
```
Zienkiewicz–Zhu patch recovery [!cite](bibtex_key)
```
replace `bibtex_key` with the actual key.

ChengHauYang replied:
I tried to fix this in the newest commit.

ChengHauYang committed:
update citation format

ChengHauYang committed:
Refactor NodalPatchRecoveryBase coefficients functions and fix comment style

- Renamed functions for clarity:
  - getCoefficientsNoCache->getCoefficients (const version)
  - getCoefficients→getCachedCoefficients (cached version)
  - Updated calls in NPRBase and ESMBase accordingly
- Adjusted comments to follow reviewer feedback:
  - API descriptions kept as Doxygen style (///)
  - implementation notes as //

ChengHauYang committed:
try to fix equation rendering in html

bwspenc commented on framework/doc/content/source/meshmodifiers/ElementSubdomainModifier.md:
```diff
@@ -96,4 +96,56 @@ Reinitialization can be further restricted by setting the parameter `old_subdoma
+The second strategy aims to address these challenges. The key idea is to initialize the solution on the *updated active* domain leveraging the solution information from the *stationary active* domain. In the current implementation, we adopt the Zienkiewicz–Zhu patch recovery technique to initialize the solution field for the DoFs on *updated active* elements.
+
+The initialization algorithm starts with constructing a patch of elements from the *stationary active* domain. By definition, solution to the variable of interest already exists on the patch. The solution is first interpolated onto the quadrature points on the *stationary active* domain, and a polynomial is fitted against the interpolated variable values using a least-squares fit. A complete set  ...
+
```
A lot more of them are rendering correctly, but there are still some issues. To resolve this you'll probably want to build the documentation locally and look at it.

ChengHauYang replied:
Thank you, @bwspenc! I have fixed the improper rendering in the newest commit (I forced pushed it). Could you help me double-check whether the newest one (https://mooseframework.inl.gov/docs/PRs/30630/site/source/meshmodifiers/ElementSubdomainModifier.html) looks good for you? Thanks!

ChengHauYang committed:
Refactor NPRBase: rename gatherSendList to be gatherRequestList for clarity, improve docstrings, and mark addToQuery as const

ChengHauYang commented on framework/include/meshmodifiers/ElementSubdomainModifierBase.h:
```diff
@@ -95,16 +130,43 @@ class ElementSubdomainModifierBase : public ElementUserObject
+  void storeOverriddenDofValues(const VariableName & var_name);
+
+  /// @brief Restore values to non-reinitialized nodes on reinitialized elements
+  void restoreOverriddenDofValues(const VariableName & var_name);
```
Fixed

ChengHauYang commented on framework/include/userobjects/NodalPatchRecoveryBase.h:
```diff
@@ -55,10 +99,59 @@ class NodalPatchRecoveryBase : public ElementUserObject
 
-  /// The multi-index table
+  /**
+   * Generates the multi-index table for a polynomial basis.
```
Fixed

ChengHauYang commented on framework/src/meshmodifiers/ElementSubdomainModifierBase.C:
```diff
@@ -77,6 +65,43 @@ ElementSubdomainModifierBase::validParams()
 
+  params.addParam<std::vector<VariableName>>(
+      "reinitialize_variables", {}, "Which variables to reinitialize when subdomain changes.");
+  MooseEnum reinit_strategy("IC POLYNOMIAL_NEIGHBOR POLYNOMIAL_WHOLE POLYNOMIAL_NEARBY", "IC");
```
Thanks for your feedback, GiudGiud. It is a good point to proceed and polish this PR better in the future.

hugary1995 replied:
Yeah, I think this is a pretty good idea. I opened an issue so we don't lose track of it: https://github.com/idaholab/moose/issues/31368

loganharbour commented:
Broke next: https://civet.inl.gov/job/3168447/
